### ISSUE TITLE
New functionality needed (2 way observables)

### ISSUE THREAD

**Issue by ggranato92** [Tudat/tudat#201]
Hi Dominic,

As agreed together, in order to work on my thesis project I would need the new functionalities of 2 way Doppler and 2 way laser ranging observables to be added to the Tudat code in the FullDevelopmentBranch, for simulation purposes.

Thank you very much and best regards,
Giovanni

**Comment by DominicDirkx** [comment]
The N-way range observations are now available on the development branch:

https://github.com/tudat/tudat/tree/FullDevelopmentBranch

At least 2-way Doppler will be added soon, possibly also generalized to N-way.

**Comment by ggranato92** [comment]
Hi Dominic,

Thank you very much for adding the N-way range observation model.

Concerning the new example you created for the least squares estimation, where can I find it? I updated the tudat but I am not able to find that file you showed me during last meeting. I was up to open the new issue about the generation of range data which sometimes are unrealistic due to the Earth obstruction.

Thanks,
Giovanni

**Comment by DominicDirkx** [comment]
The new example is on the FullDevelopmentBranch of the tudatExampleApplications:

https://github.com/Tudat/tudatExampleApplications/tree/FullDevelopmentBranch/satellitePropagatorExamples/SatellitePropagatorExamples.

**Comment by DominicDirkx** [comment]
By the way, the observation viability functionality has been added in this branch:

https://github.com/dominicdirkx/tudat/tree/ObservationViability

and now also on the development branch of tudat:

https://github.com/tudat/tudat/tree/FullDevelopmentBranch

I'll write an example application in the coming week to showcase the functionality. It allows checking observations according to:
- Minimum elevation angle
- Minimum viewing angle of some third body (e.g. Sun)
- Occultation of the link by some body (not relevant for you)

**Comment by ggranato92** [comment]
Thank you very much Dominic.

 I just found out my tudatExampleApplications folder is linked to the master branch and not the development one. Is there any easy and fast way to synchronize it with the FullDevelopmentBranch through the git-shell terminal?

**Comment by DominicDirkx** [comment]
Of course, it's the exacr same for tudat and tudatExampleApplications. Make sure to fetch the latest code from your remote first. Do

git remote -v

To get a list of your remote(s). The one you want is the github.com/tudat/... one. This one is probably called origin or upstream. Assuming it is upstream, do:

git fetch upstream
git checkout FullDevelopmentBranch 

And you should be on the right branch

**Comment by ggranato92** [comment]
Thank you, I have followed the steps and everything looks fine now

**Comment by DominicDirkx** [comment]
Quick update on this feature: the two-way doppler model is implemented and tested:

https://github.com/DominicDirkx/tudat/tree/TwoWayDoppler

The partial derivatives are not yet finished, but they are now included in the code. There still seems to be a small error that needs to be resolved, I think it's related to a scaling of the uplink partials when using the receiver as a reference, and the downlink partials when using the transmitter as reference.

**Comment by DominicDirkx** [comment]
Good news! The two-way Doppler partials are up and running. The unit tests pass at the same tolerances used for the other models. I've pushed the code to by TwoWayDoppler branch, and will merge it into the tudat/development branch when I've added all the comments.

One thing that I haven't yet done is try to use the two-way Doppler data in an orbit determination simulation. This will be a good case for you to try out. If you see any convergence issues that get worse when moving from one- to two-way Doppler, let me know. 

I'll keep this issue open until the code is merged into the development branch.
